Support textDocument/diagnostic specification (Pull diagnostics)

SofusA opened the pull request:
Closes #7757, if  [workspace diagnostics](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_diagnostic) is not a requirement. I have not yet found a language server which supports this capability. 

Handling of response was originally done by @woojiq in #7900. I was not able to include their git history since their branch has been deleted. 

Tested with language servers:
- `eslint` version >4.10.0
- `csharp-language-sever` version >4.12.0
- `ruby-lsp`
- `rust-analyzer` version >`2025-03-03`

Diagnostics are pulled when a document is changed for each language server with pull diagnostics feature with an async hook after a debounce period.

Diagnostics are also pulled when a document is opened, when changed to buffer and when a language server is successfully initiated. This is done without debounce.

the-mikedavis commented on helix-term/src/commands/lsp.rs:
```diff
@@ -1424,3 +1424,88 @@ fn compute_inlay_hints_for_view(
+
+pub fn pull_diagnostic_for_current_doc(editor: &Editor, jobs: &mut crate::job::Jobs) {
+    let doc = doc!(editor);
+    let Some(language_server) = doc
+        .language_servers_with_feature(LanguageServerFeature::PullDiagnostics)
+        .next()
```
It looks like this is only requesting diagnostics from the first language server to support it. We should request pull diagnostics for all language servers in this iterator - there might be multiple language servers configured that support the feature.

SofusA replied:
Yes that was not intentional

SofusA replied:
I have a hard time testing if multiple servers work, however it is working with the one language server that i am using.

the-mikedavis commented on helix-term/src/ui/editor.rs:
```diff
@@ -1071,6 +1071,7 @@ impl EditorView {
 
     pub fn handle_idle_timeout(&mut self, cx: &mut commands::Context) -> EventResult {
         commands::compute_inlay_hints_for_all_views(cx.editor, cx.jobs);
+        commands::pull_diagnostic_for_current_doc(cx.editor, cx.jobs);
```
Inlay hints were introduced before the event system that @pascalkuthe was mentioning in https://github.com/helix-editor/helix/pull/7900#issuecomment-1735911802. Pulling diagnostics should be done with an event system hook rather than here on idle timeout. I'm not sure we have a good PR to reference for how to do this yet. https://github.com/helix-editor/helix/pull/9668 and https://github.com/helix-editor/helix/pull/9638 show using hooks and #8021 introduced the event system initially and is probably the best background

SofusA replied:
Thanks for the review.

Ah i see. I hoping this was related to the dispatch of DiagnosticsDidChange event. This makes more sense. 

Does it make sense to register a hook for DocumentDidChange which handles pulling and adding diagnostics?

pascalkuthe replied:
yeah that should do it you need an async handler to make sure they are debounced. There are a decent amount of hooks now that subscribe to `DocumentDidChange` so you can take a look at how those work. A good lsp related example is probably signature help (just that in your case you don't care about the selection changing).


I also don't think this should really live under commands at all. I don't think we need to support manually pulling diagnostics and if we do we can use a  handler for that (like we do for manually requesting completions). All related code should be (self-contained) inside the hook. This allows allows adding complexity like tracking result-id across multiple requests.

In a sense the most elegant way to refactor this would be to make is so that you decouple our current diagnostic handling from `publish-diagnostic` and then essentially create an internal pulbish diagnostic notification that can either come from the lsp directly or from our pull diagnostic handler.

SofusA replied:
Thanks for clarification. 
I have pushed a refactoring with basically the same logic just with an async hook instead of idle timeout.

Let me know if i have missed something.

Now helix will only requesting diagnostics after document has changed. Is there a nice way to initiate the hook when the language server is started?

SofusA commented on rust-toolchain.toml:
```diff
@@ -1,3 +1,3 @@
 [toolchain]
 channel = "1.74.0"
-components = ["rustfmt", "rust-src", "clippy"]
+components = ["rustfmt", "rust-src", "clippy", "rust-analyzer"]
```
Yes that was not intentional. This has been reverted

archseer commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +166,12 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+
+    let tx = handlers.pull_diagnostics.clone();
+    register_hook!(move |event: &mut DocumentDidChange<'_>| {
+        if event.doc.config.load().lsp.auto_signature_help {
+            send_blocking(&tx, PullDiagnosticsEvent::Trigger);
```
Should this have it's own config option? Seems odd to tie it to signature help

the-mikedavis replied:
I think this is a copy/paste error. Instead we can check if the `event.doc` has a language server that supports pull diagnostics so that we don't even trigger the hook if the doc won't trigger pull diagnostics.

I'm thinking adding a method to Document:

```rust
pub fn has_language_server_with_feature(&self, feature: LanguageServerFeature) -> bool {
    self.language_servers_with_feature(feature).next().is_some()
}
```

and checking that on `event.doc` instead

SofusA replied:
Yes this was a copy/paste error. 

I added @the-mikedavis suggestion.

archseer commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -1,12 +1,157 @@
+fn pull_diagnostic_for_current_doc(editor: &mut Editor) {
+    let (_, doc) = current!(editor);
+
+    for language_server in doc.language_servers_with_feature(LanguageServerFeature::PullDiagnostics)
+    {
```
Could we check for a language server with a pull diagnostics feature in the hook first? That way we avoid dispatching a callback via `send_blocking` if the server doesn't support it (which will likely be the most common case)

SofusA replied:
I added `has_language_server_with_feature` as suggested by @the-mikedavis. This way only send when there is at least one language server with this feature.

the-mikedavis commented on helix-view/src/handlers/lsp.rs:
```diff
@@ -47,6 +47,10 @@ pub enum SignatureHelpEvent {
     RequestComplete { open: bool },
 }
 
+pub enum PullDiagnosticsEvent {
+    Trigger,
+}
```
Instead let's make this a struct (for now - we can always refactor as an enum later if we want to support multiple kinds of events) which carries the `DocumentId` from `event.doc.id()` on the `DocumentDidChange` event. Then in the hook you can get the document with `doc!(editor, &doc_id)` or `doc_mut!(editor, &doc_id)` for the mutable version

SofusA replied:
Nice. I added this, but it has resulted in document_id being copied and passed around. 
But it did not make sense to pass a mutable reference to document.

SofusA commented:
Thanks again for the reviews!
I added your comments and did some refactoring. I closed the conversations, i am sorry if it is usually the reviewer who does this. 

The request is currently only being send on document changes, which makes a terrible user experience :smile: 

Is there a way to send this request when a document is opened, or should a [workspace diagnostic event](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_diagnostic) be send when the language server (which supports this feature) is started?

SofusA commented:
I did some experiments with this on roslyn language server. It might be that this language server just behaves weird (which it does on other stuff), but sending `workspace/diagnostic` after the server is initiated always resulted in and empty diagnostic list. Also with some seconds delay.

I also noticed that responses can either be a full or an related unchanged document report. At the moment, since we only listening on changes, we only get the full reports. This means that when i eg. rename something, i will not see diagnostics on related documents until i make a change.

I am not sure if the specification suggest to pull diagnostics every time a document is viewed, or if we should pull diagnostics for all open documents on changes to a document.

SofusA commented:
My plan is to send pull diagnostics request:
1. When current document is edited for all open documents, if possible where other documents has same active language server by id as edited document
1. When language server is started for all open documents
1. When a new document is opened

1 and 2 is straightforward but 3 required a new event. Overall this is a lot of code.

SofusA commented:
I have now added two events:
*Pull diagnostics for documents*
Trigger: When documents are opened
Document ids are collected in a hashset
After a short debounce all documents pull from all its language servers with pull diagnostics support.

*Pull diagnostics for language servers*
Trigger: When document are saved
Language server ids with pull diagnostics support from saved document are collected in a hashset.
After a longer debounce all language servers for all open document are pulled

Document ids are being passed around, and i do some cloning of ids. 
Also the two functions for each event does duplicate some code which could be shared. But i did have some issues with the borrow checker.

I am also not able to send the actual document in the document did open event, because i need to wait for language sever to launch and `launch_language_servers` takes ownership of the document.

Overall it appears to work well. At least better than the first implementation with idle timeout.

SofusA reviewed (commented):
I added has_language_server_with_feature as suggested by @the-mikedavis. This way only send when there is at least one language server with this feature.

lizclipse commented:
I've been trying this PR to see if it'll solve some issues using the ESLint LS from `vscode-langservers-extracted`, and I've run into a couple of bugs. Firstly, if you open a file directly using the CLI args, then the diagnostics won't show, and instead you have to close the buffer and re-open via the file picker. Secondly, and most noticable I think, if I then close _this_ buffer and try to re-open it again then I get a crash:

```
thread 'main' panicked at helix-term/src/handlers/diagnostics.rs:144:19:
no entry found for key
stack backtrace:
   0: rust_begin_unwind
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:597:5
   1: core::panicking::panic_fmt
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:72:14
   2: core::panicking::panic_display
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:168:5
   3: core::panicking::panic_str
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:152:5
   4: core::option::expect_failed
             at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/option.rs:1988:5
   5: helix_term::handlers::diagnostics::pull_diagnostics_for_documents
   6: helix_term::job::Jobs::handle_callback
   7: helix_term::application::Application::run::{{closure}}
   8: tokio::runtime::park::CachedParkThread::block_on
   9: tokio::runtime::runtime::Runtime::block_on
  10: hx::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Of note, you do have to be using `vscode-langservers-extracted@>=4.9.0` (I'm testing on 4.10.0), as that's when they changed something to stop the LS working originally, and is why I'm trying this.

SofusA commented:
Thanks for testing!

> Firstly, if you open a file directly using the CLI args, then the diagnostics won't show, and instead you have to close the buffer and re-open via the file picker. 

I see the same behaviour with roslyn language server. I *think* this is because we are requesting diagnostics before the language server is ready. I was hoping that this would be the language servers responsibility. I am not sure what to do about this.

> Secondly, and most noticable I think, if I then close this buffer and try to re-open it again then I get a crash

I *fixed* this by skipping the diagnostics request if the document id is not in the editor. 
I am not sure if this is the best solution.

lizclipse commented:
It looks like the crash is fixed now, thank you. I'd be curious on why the diagnostic request is trying to happen on a missing document ID (is there a memory leak somewhere?), but a little defensive programming didn't hurt anyone.

I've been testing this PR against v4.10.0 and so far it's been working (unlike the master branch, since eslint needs this now I think), although I am still getting the issue where it doesn't pick up the first open file. With my testing just now, it seems that it also doesn't pick up if you open helix blank and then open a file via the picker, you then have to re-open the file (or make an edit) before it starts showing anything. IMO I think functionally it's fine for this PR, but it may be worth opening an issue afterwards as a follow-up to at least document that this is known dodgey behaviour.

SofusA commented:
I looked into it, and i am not sending the pull diagnostics request, because of a race between the event, and the launching the language server (and registering capabilities). The event wins :smile:  
In sequent events (open or edit) the language server will be initiated and capabilities are registered. 

I have not found a good solution for this, which i find very annoying.

> It looks like the crash is fixed now, thank you. I'd be curious on why the diagnostic request is trying to happen on a missing document ID (is there a memory leak somewhere?), but a little defensive programming didn't hurt anyone.

I am not completely sure why. I think part of it was because i was doing this asynchronous, where i collected all document_ids in a hashset, and then send all notifications after a debounce. I changed this to a synchronous hook instead.

SofusA commented:
I did some experiments with `diagnostics refresh`, but it does not look like the language servers that I use send this notification. I was hoping this would be sent when the language server is ready to receive pull diagnostic requests.

SofusA commented:
I have started to implement workspace diagnostics, but the language servers that i use don't responds to this request. Eslint returns an error and roslyn returns an empty array. So not much to work with here.

I also think that this will be difficult to get to work in helix, because we would need the diagnostics picker to await the async request before displaying, or we would need a way to refresh diagnostics in the picker when the request has finished. 
So i am leaving this feature.

I got refresh diagnostics working though

ssiltanen commented:
> eslint version 4.10.0
> Eslint returns an error

Have you tried building a newer version of [Microsoft/vscode-eslint](https://github.com/microsoft/vscode-eslint/releases)? Since you referred to version `4.10.0`, I imagine you are using the build from [hrsh7th/vscode-langservers-extracted](https://github.com/hrsh7th/vscode-langservers-extracted), which took a non-release commit from main at its build time.

SofusA commented:
> > eslint version 4.10.0
> > Eslint returns an error
> 
> Have you tried building a newer version of [Microsoft/vscode-eslint](https://github.com/microsoft/vscode-eslint/releases)? Since you referred to version `4.10.0`, I imagine you are using the build from [hrsh7th/vscode-langservers-extracted](https://github.com/hrsh7th/vscode-langservers-extracted), which took a non-release commit from main at its build time.

No i have only tried `vscode-langservers-extracted` from `nixpkgs`. 

But this would require some refactoring of the diagnostics picker, and i think it should be in a separate pull request

the-mikedavis commented:
See the symbol picker for an example of a picker that waits on an LSP request. The workspace symbol picker is maybe also a good example but it's a sort of "dynamic picker" that re-requests symbols as you type with follow-up requests to the language server.

the-mikedavis commented on helix-lsp/src/lib.rs:
```diff
@@ -614,6 +614,7 @@ pub enum Notification {
     ShowMessage(lsp::ShowMessageParams),
     LogMessage(lsp::LogMessageParams),
     ProgressMessage(lsp::ProgressParams),
+    WorkspaceDiagnosticRefresh,
```
It looks like this is actually a request rather than a notification in the spec, so this should live under `MethodCall`. In `application.rs` the handler for this branch should move down to where we handle `Call::MethodCall`s and return `Ok(serde_json::Value::Null)` I believe

SofusA replied:
You are right. I have moved it to `MethodCall`

the-mikedavis commented on helix-view/src/editor.rs:
```diff
@@ -2154,6 +2160,92 @@ impl Editor {
+        document_version: Option<i32>,
+        result_id: Option<String>,
+    ) {
+        let doc = self.documents.values_mut()
+                            .find(|doc| doc.uri().is_some_and(|u| u == uri))
+                            .filter(|doc| {
+                                if let Some(version) = document_version{
+                                    if version != doc.version() {
+                                        log::info!("Version ({version}) is out of date for {uri:?} (expected ({}), dropping PublishDiagnostic notification", doc.version());
+                                        return false;
+                                    }
+                                }
+                                true
+                            });
+
+        if let Some(doc) = doc {
```
While we're moving this block let's take the opportunity to refactor some of the code here:

```rust
let Some(doc) = self
    .documents
    .values_mut()
    .find(|doc| doc.uri().is_some_and(|u| u == uri))
else {
    return;
};
if let Some(version) = document_version {
    if version != doc.version() {
        log::info!("Version ({version}) is out of date for {uri:?} (expected ({}), dropping PublishDiagnostic notification", doc.version());
        return;
    }
}
```

That eliminates the extra nesting from `if let Some(doc) = doc` and IMO the `find` + `filter` is a little hard to follow in its current form.

SofusA replied:
Agree. This is much better!

the-mikedavis commented on helix-term/src/application.rs:
```diff
@@ -735,9 +743,11 @@ impl Application {
                                 language_id,
                             ));
+
+                            helix_event::dispatch(helix_view::events::DocumentDidOpen { doc });
```
Ideally we should only be dispatching the internal DocumentDidOpen when we initially open a buffer. Here instead of using `DocumentDidOpen` we should be able to use `pull_diagnostics_for_document`

SofusA replied:
Changed to use `pull_diagnostics_for_document` directly

the-mikedavis commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +32,201 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+
+#[derive(Debug)]
+pub(super) struct PullDiagnosticsHandler {
+    document_id: Option<DocumentId>,
```
It should be possible for multiple documents to change while the changes are debounced. (For example the language server might send edits for multiple documents for an LSP rename request.) Instead of tracking one document I believe we need a HashSet/BTreeSet of document IDs to request

SofusA replied:
I did not think of that scenario. 
I chose to use `HashSet`. The dataset will be very small and short lived so i am not sure what would be the best.

SofusA commented:
> See the symbol picker for an example of a picker that waits on an LSP request. The workspace symbol picker is maybe also a good example but it's a sort of "dynamic picker" that re-requests symbols as you type with follow-up requests to the language server.

Okay it does not look that difficult. I have however not been able to get response from `roslyn language server`. I will try again with a newer version of `eslint language server` next week

SofusA commented:
> I ported out the change to move the diagnostic handling from Application to Editor to master in [62625ed](https://github.com/helix-editor/helix/commit/62625eda460c7c5e7cbd76db92d2d141aa2f9650) and rebased this.
> 
> I dropped a change though in https://github.com/SofusA/helix-pull-diagnostics/blob/3304a54ed45b42e5b3819fb8dd165386e854b3a7/helix-view/src/editor.rs#L2236-L2242: that was discarding diagnostics that came in without an open document. For some language servers that's ok since they only ever emit diagnostics for the opened documents but others like rust-analyzer can send documents for the whole workspace. For example if you change a function's name in one file, rust-analyzer should send diagnostics for all other files in the workspace that need to update the name.

Awesome. Thanks! 

Is there anything I could do to help complete this pull request? I am a little lost in the process.

the-mikedavis commented:
For now no - I need to go throgh the spec and sit with the changes for a while and then I'll leave some review. I'd also like @pascalkuthe's review on the hook/event stuff but he's quite busy with work at the moment so in the meantime I'd like to minimize the changes with master so it's easy to rebase.

SofusA commented:
> For now no - I need to go throgh the spec and sit with the changes for a while and then I'll leave some review. I'd also like @pascalkuthe's review on the hook/event stuff but he's quite busy with work at the moment so in the meantime I'd like to minimize the changes with master so it's easy to rebase.

Great :+1: 

I think found an issue.

At work, we have recently added `Sonar` and `StyleCop` to our large dotnet project. This has made the diagnostics requests slow. So now diagnostics come in one by one in a rather annoying way. See attached video. I can only reproduce in with these diagnostics providers in a bigger dotnet codebase.

It don't think it is possible to cancel a pull diagnostic request, so maybe this just indicates that the debounce should be higher. Maybe 250 ms.

[Screencast From 2025-02-10 15-27-27.webm](https://github.com/user-attachments/assets/0fd916ec-2751-4876-886c-dd75120c188e)

Do you want me to  keep rebaseing and tuning the debounce value on this branch?

the-mikedavis commented:
Yeah feel free to push/rebase/force-push this branch as you see fit - I only force-pushed to resolve the conflicts and keep it to one commit as you had been doing

SofusA commented:
I noticed, that latest rust-analyzer does provide pull diagnostic capabilities. *But* only some diagnostics are included in pull diagnostic responses. This partial diagnostic is then replacing all the diagnostics from the publish diagnostics.

I am not sure what to do here, and it should be handled before merge. 
Any ideas?
My current best solution would be to store pull and push diagnostics separately, and render a joined list of diagnostics.

SofusA commented:
I managed to solve this by changing `DiagnosticProvider` to an enum. Coincidentally, there was already a todo for this. I use this enum instead of `LanguageServerId` when something is related to diagnostics from a server.

Now, a language server can provide both types of diagnostics. Both rust-analyzer and the C# language server work :tada:

SofusA commented:
> It don't think it is possible to cancel a pull diagnostic request, so maybe this just indicates that the debounce should be higher. Maybe 250 ms.

I solved this issue by pull diagnostics on the first change, and then increase the debounce to 500 ms.
This both fixes the issue and the first diagnostic is now rendered 120 ms faster.

The debounce could easily be lower, but the perceived responsiveness is fine at 500.

the-mikedavis commented:
There's some extra context in the LSP issue tracker about whether you can support push and pull diagnostics simultaneously https://github.com/microsoft/language-server-protocol/issues/1743. It sounds like rust-analyzer's behavior currently is a bug. It's also mentioned in comments after the pull diagnostic PR was merged: https://github.com/rust-lang/rust-analyzer/pull/18404#issuecomment-2525047165

the-mikedavis commented on helix-core/src/diagnostic.rs:
```diff
@@ -50,8 +50,21 @@ pub struct Diagnostic {
-pub type DiagnosticProvider = LanguageServerId;
+// TODO turn this into a feature flag when lsp becomes optional
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum DiagnosticProvider {
+    PublishDiagnosticProvider(LanguageServerId),
+    PullDiagnosticProvider(LanguageServerId),
+}
```
From the discussions I've read it sounds like this is meant to be done with the `identifier` value included on the server capabilities. R-a should add one in a future version https://github.com/rust-lang/rust-analyzer/pull/19266

So instead what I'm thinking for this enum is

```rust
pub enum DiagnosticProvider {
    Lsp {
        server_id: LanguageServerId,
        identifier: Option<String>,
    },
    // In the future, other non-LSP providers like spell checking go here...
}
```

where pull diagnostics would use that field from the capabilities and push diagnostics would always be `None`. (At least for now - the LSP spec could add an identifier to push diagnostics in the future.)

SofusA replied:
I gave this in another branch [here](https://github.com/helix-editor/helix/commit/100b9c3e7226bbae7a58cdc788796466a5b352f2)

I cannot `Copy` with the identifier string. So i had to clone the diagnostic provider approx. 10 times. Is this fine?

SofusA replied:
I have added this change now in a separate commit. 
Since i can no longer derive copy, i have cloned  `DiagnosticProvider` a bit, and added some implementation for helping comparing. 

This also breaks rust-analyzer for versions before `2025-03-03`. 
Nice job with the identifier btw.

Let me know if you think this should be in a prerequisite pull request.

the-mikedavis replied:
I'm a little split on this. We could have separate enum members

```rust
pub enum DiagnosticProvider {
    LspPush(LanguageServerId),
    LspPull {
        server_id: LanguageServerId,
        identifier: Option<String>,
    },
    // ...
}
```

to ensure they are kept separate. The spec is unfortunately not clear on how this should be treated. I _think_ in the long run we can expect the spec to say that the current design is more correct (just one enum variant shared by push and pull).

Since `rust-analyzer` is the only server to use both push and pull simultaneously (to my knowledge), I guess for now I think it's ok to treat rust-analyzer versions between the introduction of pull diagnostics and 2025-03-03 as broken for pull diagnostics (i.e. keep the single enum variant as it is right now). Pull diagnostics can always be turned off in config anyways.

SofusA replied:
I would personally prefer the current design. It _seems_ more correct. 
Especially if we conciser older versions of rust-analyzer as broken.
But it should be communicated that users needs to update.

the-mikedavis commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +32,194 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+                })
+                .await
+            }
+            Err(err) => log::error!("Pull diagnostic request failed: {err}"),
```
I believe we should also handle the error case here and look for `ServerCancelled` and retry based on the cancellation data: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticServerCancellationData

Lack of cancel+retry support for this seems to have bit the neovim+rust-analyzer combination lately: https://github.com/neovim/neovim/issues/30985 / https://github.com/neovim/neovim/pull/31345. I can reproduce getting some error logs from this line about the server cancelling the request when modifying a document shortly after opening a rust file (starting rust-analyzer)

SofusA replied:
I added parsing and retry after 500 sec delay.

the-mikedavis commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +32,194 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+                provider,
+            );
+        }
+        lsp::DocumentDiagnosticReport::Unchanged(report) => {
+            doc.previous_diagnostic_id =
+                Some(report.unchanged_document_diagnostic_report.result_id);
+
+            handle_document_diagnostic_report_kind(
+                editor,
+                document_id,
+                report.related_documents,
+                provider,
+            );
+        }
+    }
+}
+
+fn handle_document_diagnostic_report_kind(
+    editor: &mut Editor,
+    document_id: DocumentId,
+    report: Option<HashMap<lsp::Url, lsp::DocumentDiagnosticReportKind>>,
+    provider: DiagnosticProvider,
+) {
+    for (url, report) in report.into_iter().flatten() {
+        match report {
+            lsp::DocumentDiagnosticReportKind::Full(report) => {
+                let Ok(uri) = Uri::try_from(url) else {
+                    return;
+                };
+
+                editor.handle_lsp_diagnostics(provider, uri, None, report.items, report.result_id);
+            }
+            lsp::DocumentDiagnosticReportKind::Unchanged(report) => {
+                let Some(doc) = editor.document_mut(document_id) else {
+                    return;
+                };
+                doc.previous_diagnostic_id = Some(report.result_id);
+            }
+        }
+    }
```
I think we can eliminate the extra helper function here and make this a bit shorter

```rust
fn handle_pull_diagnostics_response(
    editor: &mut Editor,
    response: lsp::DocumentDiagnosticReport,
    provider: DiagnosticProvider,
    uri: Uri,
    document_id: DocumentId,
) {
    let (result_id, related_documents) = match response {
        lsp::DocumentDiagnosticReport::Full(report) => {
            editor.handle_lsp_diagnostics(
                provider,
                uri,
                None,
                report.full_document_diagnostic_report.items,
            );

            (
                report.full_document_diagnostic_report.result_id,
                report.related_documents,
            )
        }
        lsp::DocumentDiagnosticReport::Unchanged(report) => (
            Some(report.unchanged_document_diagnostic_report.result_id),
            report.related_documents,
        ),
    };

    if let Some(doc) = editor.document_mut(document_id) {
        doc.previous_diagnostic_id = result_id;
    };

    for (url, report) in related_documents.into_iter().flatten() {
        let result_id = match report {
            lsp::DocumentDiagnosticReportKind::Full(report) => {
                let Ok(uri) = Uri::try_from(url) else {
                    continue;
                };

                editor.handle_lsp_diagnostics(provider, uri, None, report.items);
                report.result_id
            }
            lsp::DocumentDiagnosticReportKind::Unchanged(report) => Some(report.result_id),
        };

        if let Some(doc) = editor.document_mut(document_id) {
            doc.previous_diagnostic_id = result_id;
        }
    }
}
```

I think we should also try to apply as much of the report (continue rather than return).

I also changed the signature of `Editor::handle_lsp_diagnostics` here so that the calling code here sets the `result_id` rather than passing it. I think that makes the option handling a little cleaner and we don't need to worry about the meaning of this field for push diagnostics.

Instead of passing the result_id though I believe we will need to pass the `identifier` (or at least the `DiagnosticProvider`).

SofusA replied:
Added :+1:

the-mikedavis commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +32,194 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+    document_ids: HashSet<DocumentId>,
+}
+
+impl PullDiagnosticsHandler {
+    pub fn new() -> PullDiagnosticsHandler {
+        PullDiagnosticsHandler {
+            document_ids: [].into(),
+        }
+    }
+}
+
+impl helix_event::AsyncHook for PullDiagnosticsHandler {
+    type Event = PullDiagnosticsEvent;
+
+    fn handle_event(
+        &mut self,
+        event: Self::Event,
+        existing_debounce: Option<tokio::time::Instant>,
+    ) -> Option<tokio::time::Instant> {
+        if existing_debounce.is_none() {
+            dispatch_pull_diagnostic_for_document(event.document_id);
+        }
+
+        self.document_ids.insert(event.document_id);
+        Some(Instant::now() + Duration::from_millis(500))
+    }
+
+    fn finish_debounce(&mut self) {
+        for document_id in self.document_ids.clone() {
+            dispatch_pull_diagnostic_for_document(document_id);
+        }
+    }
+}
```
For the handler I think we will want some more nuanced logic. There are some tips at the bottom of the "Diagnostics Refresh" section in the spec that I think are meant to apply to pull diagnostics generally and are just out of place in the spec at the moment:

> Implementation Considerations
> 
> Generally the language server specification doesn’t enforce any specific client implementation since those usually depend on how the client UI behaves. However since diagnostics can be provided on a document and workspace level here are some tips:
> 
> * a client should pull actively for the document the users types in.
> * if the server signals inter file dependencies a client should also pull for visible documents to ensure accurate diagnostics. However the pull should happen less frequently.
> * if the server signals workspace pull support a client should also pull for workspace diagnostics. It is recommended for clients to implement partial result progress for the workspace pull to allow servers to keep the request open for a long time. If a server closes a workspace diagnostic pull request the client should re-trigger the request.

So the behavior should depend on whether the server declares that the language has inter-file dependencies. For language servers with inter-file dependencies we want to query for all "visible" documents attached to that language server. And for languages without inter-file dependencies we can request diagnostics only when that document changes, and we can pull on a lower debounce.

As for what is a "visible" document I think it's meant to be left up to the editor. In our case I think it's fine to query any open document (so we could look at `Editor`'s documents instead of holding a hashset here).

SofusA replied:
> So the behavior should depend on whether the server declares that the language has inter-file dependencies. For language servers with inter-file dependencies we want to query for all "visible" documents attached to that language server. And for languages without inter-file dependencies we can request diagnostics only when that document changes, and we can pull on a lower debounce.

I think this would require two separate async hooks: One for servers with and without interfile dependencies. Is this something we want?

> As for what is a "visible" document I think it's meant to be left up to the editor. In our case I think it's fine to query any open document (so we could look at `Editor`'s documents instead of holding a hashset here).

I think this requires that i add `Editor` to the `DocumentDidChange` and `DocumentDidOpen` in `helix-view/src/events.rs`. Not that this is a problem, it will just require a lot of changes

the-mikedavis replied:
Adding editor to DocumentDidOpen shouldn't be a big deal but adding it to DocumentDidChange will not be possible for now. (Eventually it should be but we will need to move `Document::apply` from `Document` to `Editor`, which will solve other rough edges not related to this PR.)

Alternatively I think it should be possible to dispatch a job in `finish_debounce` which looks at the editor and pulls for `editor.documents()` or `editor.tree.views().map(|(view, _)| view.doc)`.

I think it should also be possible to use only one hook and still have two different timeouts. It might mean adding a field to `PullDiagnosticHandler` where the `Instant` is stored for the longer debounce time. Having multiple hooks is not a big deal though, they should be very cheap to spawn/run.

SofusA replied:
I have added a job which looks at editor, and i have have added a separate hook for pull diagnostics for diagnostic providers without inter-file dependencies.

SofusA replied:
I added a separate timeout to the hook. 

I also removed the hashset and replaced it with querying diagnostics for all open documents and their language servers.

SofusA commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +32,194 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+                    pull_diagnostics_for_document(doc, language_server);
+                }
+            })
+        }
```
Added :+1:

SofusA commented:
I have tried to address comments in separate commits.
Some changes has made this pull request large. I am sorry about that :smile:

the-mikedavis commented:
I ported out the change to refactor the diagnostic provider as an enum and the change to add `DocumentDidOpen` (as well as some others) and gave this branch a rebase: https://github.com/helix-editor/helix/compare/2d4c2a170cab5f625139b829b2d3f00b303a8550...6da1a79d80d90d13cb4821362a52fa0d6b9bdee6

Those changes will be necessary for some other patches like #11660 and DocumentDidOpen I believe was already needed by a few open PRs.

W.r.t. rust-analyzer, one option to limit the impact of the broken versions is to turn off pull diagnostics for a while by default:

```toml
[[languages]]
name = "rust"
# ...
language-servers = [{ name = "rust-analyzer", except-features = ["pull-diagnostics"] }]
# ...
```

Then once 2025-03-03 gets older we can eventually go back to using them.

the-mikedavis commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +33,265 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+
+fn handle_pull_diagnostics_response(
+    editor: &mut Editor,
+    response: lsp::DocumentDiagnosticReport,
```
It looks like the return value for the future should be `lsp::DocumentDiagnosticReportResult`: https://github.com/helix-editor/helix/blob/6da1a79d80d90d13cb4821362a52fa0d6b9bdee6/helix-lsp-types/src/request.rs#L900

(It decodes correctly as `lsp::DocumentDiagnosticReport` because the result one is an untagged enum that can be `lsp::DocumentDiagnosticReport`)

Ideally we would do the decoding in the client. This comes up fairly often actually so I might push a change to the LSP client code to always eagerly decode - there isn't really any point in doing it lazily and it can lead to type mixups like this

sandiks commented:
Hi, is it possible integrate this lang server with zeditor?

SofusA commented:
> Hi, is it possible integrate this lang server with zed editor?

This is a pull request for implementing a protocol in Helix editor. 

If you are referring to the [csharp language server](https://github.com/SofusA/csharp-language-server) that I use (and is depending on this or for helix support), please add an issue there instead 😊

jadestrong commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +33,267 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+            lsp::DocumentDiagnosticReportKind::Unchanged(report) => Some(report.result_id),
+        };
+
+        if let Some(doc) = editor.document_mut(document_id) {
```
It seems there might be an issue here. The code updates `doc.previous_diagnostic_id` for the `document_id`-associated document, but this section is iterating over `related_documents`, which are identified by `url`. Instead of using `document_id`, it may be more appropriate to fetch the document corresponding to the `url` and update its `previous_diagnostic_id`.

SofusA replied:
Good catch. I have changed it to `editor.document_by_path_mut(url.path())`.
Unfortunately, none of the language servers that i use for testing (c# and rust-analyzer) don't provide related documents, so i am unable to test this.

Maybe it would make sense to leave out this part for now and wait for language servers to implement this?
What do you think @the-mikedavis ?

SofusA replied:
I have added a commit that removes support for related document changes. I dont like code that i cant justify :smile: 
This can be added when we have a language server that provides this feature

pascalkuthe commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +33,243 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+        event: Self::Event,
+        timeout: Option<tokio::time::Instant>,
+    ) -> Option<tokio::time::Instant> {
+        if timeout.is_none() {
```
the logic here does not make sense to me. The no_inter_file_dependency timeout does not really work. What I would expect is the following:

* have a debuounce of 125ms until we refresh diagnostics for the current document
* if there are lsps with inter-file dependencies then refresh all open documents registered with one of the inter-file LSPs active for the current document.

This would be structured as the following:

* change event triggers a short debounce (125ms) and at the end requests diagnostic locally
* when the local debounces finishes and the set of active language servers with inter-file dependencies is not-empy, we send another event (containing the set of active lsp with inter-file dependencies and the active document) to a second handler with a larger debounce (500ms?) that will request diagnostics for all documents for those lsps (and exclude the document that orginated the request)

pascalkuthe replied:
also: if there is a previous request running we must either cancel and discard it or wait for it to complete. Otherwise the previous result id will not be correct

pascalkuthe replied:
and in this handler we likely only want to pull for *visible* documents.

SofusA replied:
Good points.
I have made some changes:
- there are now two handlers: 
    - one for the edited document
    - one for all visible documents with language servers with interfile dependency capability
- previous running requests are cancelled
    - the language server client holds a list of ongoing requests for documents
    - when new diagnostics are requested all ongoing requests for that document is cancelled

I am not satisfied with the way that the language server return a future and the progress token, and it is then then callers responsibility to mark the request as done. I think this could be handled inside the language server client. I will give it a try.

But is this roughly the way you expect this to work?

pascalkuthe commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +33,243 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+    job::dispatch_blocking(move |editor, _| {
+        let documents = editor.documents.values();
+
+        for document in documents {
```
if the server supports it we should use a workspace diagnostic request when pulling all documents.

the-mikedavis commented:
Pascal and I can take the branch forward as we find time to work on it. I have some changes locally I was working on to track the ongoing requests anyways. We wouldn't mind getting something relatively simple in as a first step but there are some changes we'll want to make to ensure we don't spam the language server or allow duplicate in-flight requests

SofusA commented:
> Pascal and I can take the branch forward as we find time to work on it. I have some changes locally I was working on to track the ongoing requests anyways. We wouldn't mind getting something relatively simple in as a first step but there are some changes we'll want to make to ensure we don't spam the language server or allow duplicate in-flight requests

Thanks. I appreciate your work :smil

SofusA commented:
I will have some time to work on this again. But i could need some guidance for what need to be done before this is ready for review (or merge :smile: ). 

My current solution for cancelling on-going requests has been working very well, at least in C#. But it requires the consumer to mark the requests as done, which i don't find ideal. 
I am curious for your implementation for tracking requests. 

I do understand if you are busy or other pull requests has higher priority.

kdestin commented:
I've been running this PR for a couple weeks, and it's been really nice to be able to use https://github.com/SofusA/csharp-language-server/ for C#.

But I've been _sometimes_ running into a bug when I try to open multiple files at once from the commandline.


---

# Description

When opening **two or more** files from the commandline (`hx file1 file2`), helix will sometimes hang and the editor will never display the files.

https://github.com/user-attachments/assets/e3cbbe36-eda6-4634-a2ee-f0b10b1cebd4

# Environment

* Helix: ([SofusA/helix-pull-diagnostics branch: pull-diagnostics (cabced63)](https://github.com/SofusA/helix-pull-diagnostics/tree/pull-diagnostics))
* OS: Ubuntu 22.04

# Steps to Reproduce


1. Install helix from SofusA/helix-pull-diagnostics
2. Prepare files

   ```
   .
   ├── file1.py
   └── file2.py
   ```

  <details>

  <summary> file1.py (empty file)</summary>

  ```python
  ```
  </details>

  <details>

  <summary> file2.py</summary>

  ```python
  class Foo:
      def bar(self):
          pass

      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
      # _______________________________________________________________________________________________________________
  ```
  </details>

3. Install a language server (either is fine)

    * [ty](https://github.com/astral-sh/ty) (If uv is installed, this can be done with `uv tool install ty`)
    * [ruff](https://github.com/astral-sh/ruff) (If uv is installed, this can be done with `uv tool install ruff`)

4. Open both files

   ```shell
   hx file1.py file2.py
   ```

# Expected Result

hx should open both buffers, setting `file1.py` as the active buffer.

# Observed Result

hx hangs, never showing the editor, only exiting after sending SIGINT (ctrl-c).

# Miscellaneous

I wasn't able to root cause this, but here's a couple things I noticed while making a minimal-ish repro + using this PR:

* Invocation needs atleast 2 files (`hx file1 file2 file3`)
  * I've never run into this issue just trying to open 1 file
* Only reproducible when there's a language server available for the files
  * If you run through the repro without installing a language server, `hx` won't hang
  * As far as I can tell, this isn't just limited to `ty` and `hx`. I've run into this
    with C# while using https://github.com/SofusA/csharp-language-server/.
* File contents (when opening 2 files):
  * The contents of the first file seems to be irrelevant (can be empty).
  * The second file needs to be "sufficiently large" to trigger the hang
    * I couldn't determine a consistent size threshold, since this seemed to change based on the actual contents of the file.
    * This might be the reason why the hang isn't 100% reproducible when trying to open two or more files.

SofusA commented:
> I've been running this PR for a couple weeks, and it's been really nice to be able to use https://github.com/SofusA/csharp-language-server/ for C#.
> 
> But I've been _sometimes_ running into a bug when I try to open multiple files at once from the commandline.
> # Description
> 
> When opening **two or more** files from the commandline (`hx file1 file2`), helix will sometimes hang and the editor will never display the files.
> Screencast.from.09-18-2025.06.19.57.PM.mp4
> # Environment
> 
>     * Helix: ([SofusA/helix-pull-diagnostics branch: pull-diagnostics (cabced63)](https://github.com/SofusA/helix-pull-diagnostics/tree/pull-diagnostics))
> 
>     * OS: Ubuntu 22.04
> 
> 
> # Steps to Reproduce
> 
>     1. Install helix from SofusA/helix-pull-diagnostics
> 
>     2. Prepare files
>        ```
>        .
>        ├── file1.py
>        └── file2.py
>        ```
> 
> 
> file1.py (empty file)
> 
> file2.py
> 
>     3. Install a language server (either is fine)
>        
>        * [ty](https://github.com/astral-sh/ty) (If uv is installed, this can be done with `uv tool install ty`)
>        * [ruff](https://github.com/astral-sh/ruff) (If uv is installed, this can be done with `uv tool install ruff`)
> 
>     4. Open both files
>        ```shell
>        hx file1.py file2.py
>        ```
> 
> 
> # Expected Result
> 
> hx should open both buffers, setting `file1.py` as the active buffer.
> # Observed Result
> 
> hx hangs, never showing the editor, only exiting after sending SIGINT (ctrl-c).
> # Miscellaneous
> 
> I wasn't able to root cause this, but here's a couple things I noticed while making a minimal-ish repro + using this PR:
> 
>     * Invocation needs atleast 2 files (`hx file1 file2 file3`)
>       
>       * I've never run into this issue just trying to open 1 file
> 
>     * Only reproducible when there's a language server available for the files
>       
>       * If you run through the repro without installing a language server, `hx` won't hang
>       * As far as I can tell, this isn't just limited to `ty` and `hx`. I've run into this
>         with C# while using https://github.com/SofusA/csharp-language-server/.
> 
>     * File contents (when opening 2 files):
>       
>       * The contents of the first file seems to be irrelevant (can be empty).
>       * The second file needs to be "sufficiently large" to trigger the hang
>         
>         * I couldn't determine a consistent size threshold, since this seemed to change based on the actual contents of the file.
>         * This might be the reason why the hang isn't 100% reproducible when trying to open two or more files.

Thanks for the detailed report. It sounds frustrating. 
I am not able to reproduce this with C# and SofusA/csharp-language-server on fedora 42 with this branch merged with `master`.

Does the hang also occur on `master`?

Can you add information about your OS, and try again on the rebased branch?

the-mikedavis commented on helix-term/src/handlers/diagnostics.rs:
```diff
@@ -21,4 +36,264 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
+impl PullDiagnosticsHandler {
+    pub fn new() -> Self {
+        PullDiagnosticsHandler {
+            document_ids: Default::default(),
```
Since this is basically `Default`, let's remove the `::new` constructor here, add `Default` to the `#[derive(..)]` for this type, and construct it with `Default::default()` or `PullDiagnosticHandler::default()` anywhere `::new` is called now

the-mikedavis commented on helix-view/src/handlers/lsp.rs:
```diff
@@ -30,6 +30,12 @@ pub enum SignatureHelpEvent {
+    pub document_id: DocumentId,
+}
+
+pub struct PullAllDocumentsDiagnosticsEvent {}
```
For this hook I think we should try to pass in the `LanguageServerId`s of language servers that declare inter-file dependencies. You might have a simpler language server (for example one checking for typos) and requesting to refresh all of its documents would be a bit wasteful. It's a bit of a hypothetical scenario though - I wouldn't consider this to be a merge blocker.

SofusA replied:
I have pushed a commit for this. 
I think the only downside, is that i have to get the language server id from client, pass this to the other function, which gets the client from the id :smile:

the-mikedavis reviewed (commented):
I think this is looking good, I only have some minor comments. At this point I'd be inclined to just merge this as-is and we can iterate on the hooks in follow-ups if needed
